### DISCUSSION TITLE
Dockerized remote dev with conda and dependency detection

### DISCUSSION THREAD

**Discussion by maresb** [pyscaffold/pyscaffold#426] [Best Practices & Workflows]
Migrated from #135 

For a current project of mine I have developed a workflow that could be of public interest.

I develop remotely and use conda-forge to manage my development environment. The code I write is then delivered to the customer via a Docker image. I have a script which parses my src/ for imports, works out the corresponding conda/pip packages in my development environment, and writes out the subset of relevant dependencies into an environment.yaml. Then in my Dockerfile I create a base conda environment and install this environment.yaml.

This workflow makes it easy for me to frequently update my development environment. I just need to run my script to update my Docker environment.

I'm leaving out and oversimplifying a bunch of details, but if people are interested I could explain some more.

**Comment by maresb** [comment]
@FlorianWilhelm writes:

Thanks for explaining your workflow @maresb. I am not sure I quite understand it. What do you mean by

> I use conda-forge to manage my development environment.

So you create a conda environment using the `conda-forge` channel? If you have your conda environment anyway why do you then need to parse your Python `src` folder for imports to automatically create an `environment.yml` from it? Do you have a tool for that as this process seems really to be quite complex. If you import for instance `pytorch` how do you know that it's only available in the `pytorch` channel?

My workflow, also for customer projects is, I start with `putup --dsproject ...`, edit the generated `environment.yml`, create the conda environment, activate it and develop. Whenever I feel like I need one more library, I install it with conda, add it to `environment.yml` and import it. I admit that there is a repetition in this workflow in the sense that I install a package and also need to add it to the `environment.yml`. Maybe a cleaner way would be to add a new library directly to `environment.yml` and call afterwards:

    conda env update -f environment.yml --prune

which will then install the library, potentially also removing libraries you don't need anymore.

**Reply by maresb** [reply]
@FlorianWilhelm good questions. 

> What do you mean by "I use conda-forge to manage my development environment."

I develop entirely remotely in the customer's cloud. I chose to build my development environment as a Docker image with [Jupyter Docker Stacks](https://github.com/jupyter/docker-stacks) as a base image. That image uses conda-forge for package management, so I install tons of additional conda-forge packages on top of that. I also define a `pip` section in my `environment.yml` for a few PyPI packages which aren't in conda-forge. (I install [code-server](https://github.com/cdr/code-server), but actually not from conda since that [leads to dependency problems](https://github.com/conda-forge/code-server-feedstock/issues/17).)

> If you have your conda environment anyway why do you then need to parse your Python src folder for imports to automatically create an environment.yml from it?

Otherwise the Docker image I deploy into production for the customer would be huge. I include the productive code with dependencies, and exclude all my exploratory data analysis tools.

> Do you have a tool for that as this process seems really to be quite complex.

I have a hand-written Python script which leverages [`pipreqs`](https://pypi.org/project/pipreqs/) and [`conda_env`](https://github.com/conda/conda/blob/master/conda_env/env.py) to prune to my `src/` requirements. I have to disambiguate the imports as either conda or pip packages. Most of the time it's easy, but I maintain by hand a list of exceptions:
```python
import_to_conda = dict(
    matplotlib="matplotlib-base",
    sklearn="scikit-learn",
    ...
)
```
It sounds really horrible, but in reality it's really not difficult. (Just a few hundred lines very heavily documented and reasonably robust code.)

Unfortunately having a hand-made script locks me in as the administrator. It would be nice to either find a more standard way of doing things or open-source my current method so as to leverage the community.

> If you import for instance pytorch how do you know that it's only available in the pytorch channel?

I typically use TF for gradient descent and PyMC3 for Bayes, so I don't have pytorch installed at the moment. [Docker stacks installs TF via pip](https://github.com/jupyter/docker-stacks/blob/master/tensorflow-notebook/Dockerfile). There is a [painful effort underway](https://github.com/conda-forge/tensorflow-feedstock) to correct the glaring omission of TF2 in conda-forge. But if I were to install pytorch, I'd figure out whatever conda commands were necessary and add them as appropriate to my development Dockerfile.

---

In summary, what I have automated for myself and which (as far as I know) PyScaffold doesn't support is:
1. Dockerization of the project
2. Automatic dependency maintenance

Given my constraint of remote development, these two items are somewhat intertwined and also particular to my case.

I'm sure this could all be done in a simpler way if I were just familiar with the right tools. :stuck_out_tongue:

**Comment by FlorianWilhelm** [comment]
@maresb, also note that the [PyScaffold DSProject extension](https://github.com/pyscaffold/pyscaffoldext-dsproject) now comes with a Dockerfile that makes containerization quite easy. Also, check out the corresponding CI test.

**Reply by maresb** [reply]
Thanks so much @FlorianWilhelm for letting me know! I'll have a look.
